### ISSUE TITLE
Clarify our merge policies vs. guidelines

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/team-compass#594]
I've found that our [team merging policies](https://jupyterhub-team-compass.readthedocs.io/en/latest/practices/merging.html) are helpful, but a bit verbose, and focused more on "what kinds of criteria should I use in reviewing" rather than "when can I merge". I think it'd be helpful if we split those docs into two sections:

- **When can a PR be merged?**. This should be short and straightforward. I'd propose: "Any PR that has approval from another JupyterHub team member can be merged by anybody, anytime." Along with "Use your best judgment about whether to intentionally leave it open a little bit longer in order to allow others to provide feedback or voice objections."
- **Considerations in a PR review**. This would be much of our current content about what kinds of stuff to focus on etc.

**Comment by minrk** [comment]
Looks great! 

> Any straightforward or controversial

Should this be 'straightforward or _un_controversial'?

**Comment by manics** [comment]
Relevant other issues:
- https://github.com/jupyterhub/team-compass/issues/465
- https://github.com/jupyterhub/team-compass/issues/466

**Comment by choldgraf** [comment]
@manics good point - I had a specific suggestion here too:

- https://github.com/jupyterhub/team-compass/issues/466#issuecomment-965815478

I can make a PR to formalize these ideas once my latest batch are resolved, I don't wanna overwhelm people with policy proposals 😅

**Comment by betatim** [comment]
I like short timelines but I think three working days is too short to get feedback from people who might be contributing in their spare time (weekend warriors). Combined with the fact that council business is hardly ever urgent, I'd go for something like ten calendar days or some such. I think that is still short enough to create a sense of urgency and purpose, but will reduce the number of times someone misses out because they took a few days off or had other hobbies that got priority for a few days.

**Comment by consideRatio** [comment]
## Merging on approval / awaiting further review

I've thought about the choice of merging vs. awaiting further review recently. I think our repositories would benefit from more maintenance capacity than we currently have available, and therefore value our existing capacity greatly. That makes me value merging directly on approval.

I see some significant benefits to merging on approval, such as:
- A faster merge helps the PR author/other contributors continue building on the work and avoid merge conflicts, but most importantly stay active as a contributor and build up a momentum to contribute.
  - As a practical positive example, I'd like to highlight [this amazing flurry of thorough PRs to KubeIngressProxy/KubeSpawner from @dolfinus](https://github.com/jupyterhub/kubespawner/pulls?q=is%3Apr+author%3Adolfinus+) that I think of as enabled by multiple maintainers review efforts. Thank you @dolfinus :heart: :tada: :sunflower: for your thorough and thohtful work!!
- Less review needed for a PR could be review effort saved for another un-reviewed PR, and could be a context switch less for the PR author that may be very attentive to notifications about the PR.
  - As a related practical example, I merged #591 myself but took on some self-merge doubts and looked at my mobile with each notification before doing it. I know that I have some mental relief when someone else merges my PR as compared to doing it myself - even though its approved. In this case, I also think @choldgraf would have been able to avoid a merge conflict from two more recent PRs.

## Policy suggestion: default to merge on approval

- We default to merging on approval
- When not mergig on approval, we clarify why not and help nudge the PR towards resolving. Examples:
  - "This LGTM, but I wasn't certain about X. Hey A or B, what do you think about X - okay to merge?"
  - "This LGTM, but I see you just added a commit, if you are happy with the PR as it is now I'd be happy to go for a merge!"
  - "This LGTM, but as it is a possibly controversial change, I'd like another maintainer approve this PR as well. A or B, what do you think?"
  - In the non-typical situation in jupyterhub/mybinder.org-deploy, where a merged PR means we deploy mybinder.org: "This LGTM! I suspect there could be issues in the continuous deployment when merging this that I'm not willing/able to handle. Could you self-merge this and check up on it?"

**Comment by sgibson91** [comment]
I really like the model of PR authors merging their own PRs after receiving a positive review. They often have the most context of what could go wrong, so I don't like to merge other people's PRs often, unless it's just docs or something really simple that I could figure out on the fly.

**Comment by consideRatio** [comment]
Besides mybinder.org-deploy, that i think we may agree to treat specially because it has continuous deployment to production which sets it apart, do you prefer the approve without merge in other jupyterhub repos as well @sgibson91?

**Comment by sgibson91** [comment]
Definitely, because I am probably less active in other repos and therefore touch the codebase less 😁

**Comment by minrk** [comment]
I agree that self-merge after approval is usually ideal for most PRs to mybinder.org. The main reasons I would approve but not merge on non-continuous-deployment repos are:

1. I don't have sufficient confidence in the repo or specific changes, and would like someone else to take final responsibility (this could be the original author)
2. there's already a review exchange going on and it looks like the ball is in the reviewer's court. I don't want to cut it off, which feels like ignoring the other reviewer's feedback, but I want to lend my support to the changes when they are deemed ready by others already involved in review
3. a review from another person has been explicitly _requested_, but not yet received. This is tricky because it's hard to distinguish a request for multiple reviewers from requesting _all_ reviewers look at it vs _any_ reviewer. Again, in this case the author knows what they meant and can merge after approval if they want.

All of which boil down to: I would like someone else to look at this first. But in most cases, I would click merge directly and I would encourage others to do the same. For most repos, if something needs fixing a follow-up PR is always okay. There's no dire consequences to something incomplete landing in main.

**Comment by manics** [comment]
I think a PR approval without comments or other action should mean the author can take the initiative to merge.

Sometimes I'm quickly browsing through PRs on my phone when I've got a spare minute or two, if I approve without commenting/merging it means I'm happy for it to be merged but I'm not sure whether the author is ready for it to be merged and I can't quickly type a meaningful comment on my phone. There's a discussion to be had over whether it's better not to review in that situation, but often these are PRs that have been unreviewed for a few days, and I think it's better to give a quick approval than to leave them waiting for even longer.

I agree with @betatim that "at least three working days" is too short for anyone working part-time on Jupyter. "At least a week" sounds better to me.

**Comment by choldgraf** [comment]
FYI I've got a PR to propose these ideas in our team compass here:

- https://github.com/jupyterhub/team-compass/pull/596

**Comment by betatim** [comment]
Thanks a lot for the comments on "why leave a review but not merge". It turns out it is far more nuanced and they make sense. Given how diverse the answers are, maybe we can start leaving comments to tell others what the review means.

Unsurprisingly the reasons are far more nuanced than my old man reaction of "if you like it, why not merge it? That is the bit that requires commit(ment) after all?" :D
